10_rest/nickolee/pkg/puppy/rest/rest.go

```diff
@@ -0,0 +1,130 @@
+	if err := render.DecodeJSON(r.Body, &pup); err != nil {
+		http.Error(w, "Bad Request: "+invalidInputMsg, http.StatusBadRequest)
+		return
+	}
+	// Actually create a new Puppy in store
+	id, err := phs.Storage.CreatePuppy(&pup)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnprocessableEntity) // if err return a 422
+		return
+	}
+	// else if all is well tell the client the puppy has been successfully created
+	render.Status(r, http.StatusCreated)
+	pup.ID = id            // tacking on the id generated by the server
+	render.JSON(w, r, pup) // confirm this is what has been created on backend
+}
+
+// Implmenting 3/4 methods in PuppyHandler interface: puppy handler for  PUT /api/puppy/{id}
+func (phs *PuppyHandlerAndStorer) handlePut(w http.ResponseWriter, r *http.Request) {
+	// parse incoming request url param
+	id, err := strconv.Atoi(chi.URLParam(r, "id")) // strip off the {id} part of the endpoint and convert to int
+	if err != nil {
+		http.Error(w, "Unprocessable Entity: "+invalidInputMsg, http.StatusBadRequest)
+		return
+	}
+
+	// check if puppy with given id exists in storage
+	_, err = phs.Storage.ReadPuppy(id)
+	if err != nil {
+		http.Error(w, notFoundMsg, http.StatusNotFound)
+		return
+	}
+
+	// now deal with the body of the PUT request
+	var pup puppy.Puppy
+	// decode json and unmarshal into var pup
+	if err := render.DecodeJSON(r.Body, &pup); err != nil {
+		http.Error(w, "Unprocessable Entity: "+invalidInputMsg, http.StatusUnprocessableEntity)
+		return
+	}
+	// Actually update corresponding Puppy in store
+	err = phs.Storage.UpdatePuppy(id, &pup)
+	if err != nil {
+		http.Error(w, "Unprocessable Entity: "+invalidInputMsg, http.StatusUnprocessableEntity) // if err return a 422
+		return
+	}
+	// else if all is well tell the client the puppy has been successfully updated
+	render.Status(r, http.StatusCreated)
+	render.JSON(w, r, pup) // confirm this is what has been created on backend
+}
+
+// Implmenting 4/4 methods in PuppyHandler interface: puppy handler for DELETE /api/puppy/{id}
+func (phs *PuppyHandlerAndStorer) handleDelete(w http.ResponseWriter, r *http.Request) {
+	// parse incoming request url param
+	id, err := strconv.Atoi(chi.URLParam(r, "id")) // strip off the {id} part of the endpoint and convert to int
+	if err != nil {
+		// if err != nil means user didn't provide proper id hence 400
+		http.Error(w, "Bad Request: "+invalidInputMsg, http.StatusBadRequest)
+		return
+	}
+
+	// actually perform delete operation
+	err = phs.Storage.DeletePuppy(id)
+	if err != nil {
+		http.Error(w, notFoundMsg, http.StatusNotFound)
+		return
+	}
+
+	// give client feedback. Opted for this rather than render.Status(r, http.StatusNoContent) as gives more feedback
+	render.JSON(w, r, "Puppy successfully deleted")
+}
+
+// SetupRoutes maps the exposed endpoints with their respective handlers
+func SetupRoutes(r chi.Router, phs PuppyHandlerAndStorer) {
+	r.Get("/api/puppy/{id}", phs.handleGet)
```

Comment:
I would suggest experimenting with subroutes
```
r.Route("/path", func(r chi.Router) {
  r.Get("/subpath", handler)
})
```

Also, usually API's served over http have a version in the route, `api/v1/path`